crates/wysiwyg/src/dom/parser/parse.rs

```diff
@@ -616,70 +893,153 @@ mod sys {
+
+        for node in nodes {
+            if node.is_line_break() {
+                line_breaks.push(Some(node.handle()));
+                next_lines.push(next_handle.clone());
             }
+            next_handle = node.handle();
         }
-        next_handle = node.handle();
-        if node.handle().depth() <= handle.depth() {
-            break;
-        }
+
+        line_breaks.push(None);
+        next_lines.push(next_handle.clone());
+
+        (line_breaks, next_lines)
+    };
+
+    // If there were no line breaks we might stop here
+    if lines.len() <= 1 // (<= 1 because lines will always contain at least the container)
+        // Code blocks require all inline content to be wrapped in a paragraph
+        && dom.lookup_node(handle).kind() != DomNodeKind::CodeBlock
+    {
+        return dom;
     }
 
-    for line_break_handle in line_break_handles {
-        let mut sub_tree =
-            dom.split_sub_tree_from(&line_break_handle, 0, handle.depth());
-        if line_break_handle.index_in_parent() > 0 {
-            // Remove line break too
-            dom.remove(&line_break_handle.prev_sibling());
+    // Create a new node to hold the processed contents if necessary
+    let new_node = match container_node.kind() {
+        ContainerNodeKind::Paragraph => None,
+        _ => Some(container_node.clone_with_new_children(vec![])),
+    };
+
+    // Remove each line from the DOM and collect it in a vector
+    let contents = {
+        let mut contents = Vec::new();
+        for (i, line_handle) in lines.iter().enumerate() {
+            let mut sub_tree =
+                dom.split_sub_tree_from(line_handle, 0, handle.depth());
+
+            if let Some(line_break_handle) = &line_breaks[i] {
+                dom.remove(line_break_handle);
+            }
+
+            group_inline_nodes(sub_tree.document_mut().remove_children())
+                .iter()
+                .rev()
+                .for_each(|n| contents.insert(0, n.clone()));
         }
-        let node =
-            DomNode::new_paragraph(sub_tree.document_mut().remove_children());
-        children.insert(0, node);
+        contents
+    };
+
+    if handle.is_root() {
+        return Dom::new(contents);
     }
 
     let needs_removal = if dom.contains(handle) {
         let block = dom.lookup_node(handle);
-        block.kind() == CodeBlock && block.is_empty()
+        block.is_empty()
     } else {
         false
     };
+
     if needs_removal {
         dom.remove(handle);
     }
 
-    dom.insert_at(handle, DomNode::new_code_block(children));
+    // Insert the processed contents back into the dom
+    match new_node {
+        Some(mut n) => {
+            n.set_handle(handle.clone());
+            n.append_children(contents);
+            dom.insert_at(handle, DomNode::Container(n));
+        }
+        None => {
+            dom.insert(handle, contents);
+        }
+    }
     dom
 }
 
+// Group consecutive inline nodes into paragraphs.
```

Comment:
Could you describe a bit the needs to group consecutive inline nodes into paragraphs?

Is it that we are removing the linebreak type and adding the requirement of inline items to be grouped in a paragraphs, to simplify the implementation and assumptions we can make?